crates/lib/src/qpu/api.rs

```diff
@@ -91,74 +72,287 @@ impl From<String> for JobId {
-        }
-    }?;
+    let mut controller_client = execution_options
+        .connection_strategy
+        .get_controller_client(client, &quantum_processor_id)
+        .await?;
 
-    controller_client
+    Ok(controller_client
         .get_controller_job_results(request)
-        .await?
+        .await
+        .map_err(GrpcClientError::RequestFailed)?
         .into_inner()
         .result
-        .ok_or_else(|| GrpcClientError::ResponseEmpty("Job Execution Results".into()))
+        .ok_or_else(|| GrpcClientError::ResponseEmpty("Job Execution Results".into()))?)
+}
+
+/// Options avaialable when executing a job on a QPU.
+///
+/// Use [`Default`] to get a reasonable set of defaults, or start with [`ExecutionOptionsBuilder`]
+/// to build a custom set of options.
+#[derive(Builder, Clone, Debug, Default, PartialEq, Eq)]
+pub struct ExecutionOptions {
+    #[doc = "The [`ConnectionStrategy`] to use to establish a connection to the QPU."]
+    connection_strategy: ConnectionStrategy,
+}
+
+impl ExecutionOptions {
+    /// Get the [`ConnectionStrategy`]
+    #[must_use]
+    pub fn connection_strategy(&self) -> &ConnectionStrategy {
+        &self.connection_strategy
+    }
+}
+
+/// The connection strategy to use when submitting and retrieving jobs from a QPU.
+#[derive(Clone, Debug, Default, PartialEq, Eq)]
+pub enum ConnectionStrategy {
+    /// Connect through the publicly accessible gateway.
+    #[default]
+    GatewayOnly,
+    /// Connect directly to the default endpoint, bypassing the gateway. Should only be used when you
+    /// have direct network access and an active reservation.
+    DirectAccess,
+
+    /// Connect directly to a specific endpoint using its ID.
+    EndpointId(String),
+}
+
+/// Methods that help select the right controller service client given a quantum processor ID and
+/// [`ConnectionStrategy`].
+impl ConnectionStrategy {
+    fn get_job_target(&self, quantum_processor_id: &str) -> execute_controller_job_request::Target {
+        match self {
+            Self::EndpointId(endpoint_id) => {
+                execute_controller_job_request::Target::EndpointId(endpoint_id.to_string())
+            }
+            _ => execute_controller_job_request::Target::QuantumProcessorId(
+                quantum_processor_id.to_string(),
+            ),
+        }
+    }
+
+    fn get_results_target(
+        &self,
+        quantum_processor_id: &str,
+    ) -> get_controller_job_results_request::Target {
+        match self {
+            Self::EndpointId(endpoint_id) => {
+                get_controller_job_results_request::Target::EndpointId(endpoint_id.to_string())
+            }
+            _ => get_controller_job_results_request::Target::QuantumProcessorId(
+                quantum_processor_id.to_string(),
+            ),
+        }
+    }
+
+    async fn get_controller_client(
+        &self,
+        client: &Qcs,
+        quantum_processor_id: &str,
```

Comment:
Likewise.

For all of these, if it's easier to return a `Result` that complains about needing a QPU ID, that's fair too